android/app/src/main/java/com/google/samples/cronet_sample/MainActivity.java

```diff
@@ -83,49 +87,48 @@ private void onItemsLoadComplete() {
     }
 
     private void setUpToolbar() {
-        Toolbar toolbar =  findViewById(R.id.toolbar);
+        Toolbar toolbar = findViewById(R.id.toolbar);
         setSupportActionBar(toolbar);
         getSupportActionBar().setDisplayShowTitleEnabled(false);
         ((TextView) toolbar.findViewById(R.id.title)).setText(R.string.toolbar_title);
     }
 
     /**
      * This calculates and sets on the UI the latency of loading images with Cronet.
-     * @param cronetLatency
+     *
+     * <p>This method must be thread safe as it can be called from multiple Cronet callbacks
+     * in parallel.
      */
-    public void addCronetLatency(final long cronetLatency) {
-
-        totalLatency += cronetLatency;
-        numberOfImages++;
-
-        if (numberOfImages == ImageRepository.numberOfImages()) {
-            final long averageLatency = totalLatency / numberOfImages;
+    public void onCronetImageLoadSuccessful(long requestLatencyNanos) {
+        CronetMetrics delta = new CronetMetrics(requestLatencyNanos, 1);
+        CronetMetrics newMetrics = metrics.accumulateAndGet(
+                delta,
+                (left, right) -> new CronetMetrics(
+                        left.totalLatencyNanos + right.totalLatencyNanos,
+                        left.numberOfLoadedImages + right.numberOfLoadedImages));
+
+
+        if (newMetrics.numberOfLoadedImages == Math.min(
+                getCronetApplication().imagesToLoadCeiling.get(),
+                ImageRepository.numberOfImages())) {
```

Comment:
Sorry, I'm a bit lost here. Why are we getting the min?

Reply:
See the comment on imagesToLoadCeiling in CronetApplication. It's used to limit the number of images we load so that we can demonstrate how caching works. 

Each time you click the load image button imagesToLoadCeiling is incremented by one. If someone really likes the app and manages to click the button many times, we never attempt to read more images than we have in the repository.